ufo2ir/src/toir.rs

```diff
@@ -1,19 +1,109 @@
+    designspace
+        .sources
+        .iter()
+        .map(|s| (dir.join(&s.filename), to_ir_location(&s.location)))
+        .map(|(p, l)| norad::Font::load_requested_data(&p, &datareq)
+            .map(|f| Font { 
+                path: p,
+                location: l,
+                font_info: f.font_info,
+            })
+        )
+        .collect::<Result<Vec<Font>, _>>()
+        .map_err(UfoToIrError::UfoLoadError)
+}
+
+fn extract_upem(val: NonNegativeIntegerOrFloat) -> Result<u16, UfoToIrError> {
+    let val = val.as_f64();
+    if val > u16::MAX.into() {
+        return Err(UfoToIrError::AmbiguousUpemError);
+    }
+    if val - val.trunc() > f64::EPSILON {
+        return Err(UfoToIrError::AmbiguousUpemError);
+    }
+    Ok(val as u16)
+}
+
+fn upem(fonts: &Vec<Font>) -> Result<u16, UfoToIrError> {
+    // Optional NonNegativeIntegerOrFloat for a u16 field is super awesome
+    let upem: HashSet<u16> = fonts
+        .into_iter()
+        .filter_map(|f| f.font_info.units_per_em)
+        .map(extract_upem)
+        .collect::<Result<HashSet<u16>, _>>()?;
+    if upem.len() != 1 {
+        return Err(UfoToIrError::AmbiguousUpemError);
+    }
+    Ok(*upem.iter().next().unwrap())
+}
+
+pub fn designspace_to_ir(path: impl AsRef<Path>) -> Result<(ir::Font, Vec<ir::Glyph>), UfoToIrError> {
+    let designspace =
+        DesignSpaceDocument::load(&path).map_err(UfoToIrError::DesignSpaceLoadError)?;
+    let dir = path.as_ref().parent().unwrap(); // designspace *must* exist in a directory
+
+    let axes: Vec<ir::Axis> = designspace.axes.iter().map(to_ir_axis).collect();
 
-    // Someday we will return something useful! But ... not today.
-    Ok(ir_axes)
+    // TODO: to support file-change checks we may want to do our own granular loading
+    let fonts = fonts(&designspace, dir)?;
+    let upem = upem(&fonts)?;
+
+    let mut glyphs = HashMap::<String, ir::Glyph>::new();
+    for font in fonts.iter() {
+        let glyph_dir = dir.join("glyphs");
+        for glif_file in fs::read_dir(glyph_dir).map_err(UfoToIrError::IoError)? {
+            let glif_file = glif_file.map_err(UfoToIrError::IoError)?.path();            
+            let norad_glyph = norad::Glyph::load(&glif_file).map_err(UfoToIrError::GlifLoadError)?;
+            eprintln!("{:#?} {:#?}", glif_file, norad_glyph.name());
+            let name = norad_glyph.name().as_str().to_owned();
+            if !glyphs.contains_key(&name) {
+                glyphs.insert(name.clone(), ir::Glyph {
+                    name: name.clone(),
+                    instances: Vec::new(),
+                });
+            }
+            
+            let glyph = glyphs.get_mut(&name).unwrap();
+
+            let glyph_instance = ir::GlyphInstance {
+                location: font.location.clone(),
+                width: None,
+                height: None,                
+            };
+        }
+    }
+
+    Ok((ir::Font { upem, axes }, vec![]))
```

Comment:
the assumption that one designspace file maps to only one (variable) font does no longer hold for DesignSpace version 5, where one can have multiple `<variable-fonts>`, with each covering a subset of the whole designspace, see:

https://fonttools.readthedocs.io/en/latest/designspaceLib/xml.html#variable-fonts-element

so I suppose this method should return a Vec[fontir::Font].

I don't see the purpose of returning the Vec[fontir::Glyph] separately from the Font, since normally a Font owns the Glyph. Unless you are thinking about reusing glyphs in multiple fonts......

Reply:
I think we should rename Font to something like FontInfo and not return glyphs at all. The compiler will likely require the FontInfo (basic stuff like what axes, upem, etc) be completed before it processes glyphs and will then process glyphs in parallel so we'll actually never want a "do the whole thing at once" method at all.

Reply:
renamed Font=>FontInfo in https://github.com/googlefonts/fontmake-rs/pull/4/commits/cd06f47c52d0448ecceed964aecb672a0925e87f